README.md
---
title: decr
output: github_document
---



[![Travis-CI Build Status](https://travis-ci.org/gibonet/decr.svg?branch=master)](https://travis-ci.org/gibonet/decr) [![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/decr)](http://cran.r-project.org/package=decr)

The decr package provides functions to decompose observed differences in distributional statistics of a numeric variable (y) between two groups.

An example of such analysis is the observed difference between the average wages of men and women.

The decomposition proposed here is of the Blinder (1973) and Oaxaca (1973) type, whose aim is to separate the observed difference between average wages of two groups in two components:

- One that is attributable to the different distributions of observed characteristics of the two groups;
- One that is not explainable by the different distributions of observed characteristics of the two groups.

The decompositions in decr are performed nonparametrically with the reweighting approach of Di Nardo, Fortin and Lemieux (DFL, 1996) and, for what concerns the common support, it is possible to decompose the difference of any distributional statistics (not only the mean). In the case of the difference between the averages (of y) of two groups, a decomposition in 4 components is done, as in Nopo (2008). The reweighting factors, in the common support, are estimated directly from the ratio of the relative frequencies observed in the strata for the two groups. In this sense, the estimation approach is very similar to coarsened exact matching (Iacus, King and Porro 2011), except that here it is possible to consider, if any, sampling weights.


# Installation

At the moment, `decr` is available only on github and can be installed with:

```
remotes::install_github("gibonet/decr", build_vignettes = TRUE)
```

It will be probably available on CRAN after some further testing.


# Getting started



To get started, read the "decr Basic Example" vignette, that can be opened with:

```
vignette("decr_intro", package = "decr")
```



# References

Blinder, A. S. 1973. Wage Discrimination: Reduced Form and Structural Estimates. *The Journal of Human Resources* VIII (4): 436-55.

DiNardo J., N. M. Fortin and T. Lemieux. 1996. Labor Market Institutions and the Distribution of Wages, 1973-1992: A Semiparametric Approach. *Econometrica* 64 (5): 1001-44.

Iacus S. M., G. King and G. Porro. 2011. Causal Inference without Balance Checking: Coarsened Exact Matching. *Political Analysis* 20: 1-24.

Nopo, H. 2008. Matching as a Tool to Decompose Wage Gaps. *Review of Economics and Statistics*, 90(2): 290-299.

Oaxaca, R. 1973. Male-Female Wage Differentials in Urban Labor Markets. *International Economic Review* 14 (3): 693-709.


NEWS
decr 0.0.8
----------------------------------------------------------------

* lazyeval has been substituted with rlang (tidy evaluation).


decr 0.0.7.9032
----------------------------------------------------------------

* New functions (experimental): dec_quantiles() and dec_all_(), which perform
  decompositions of quantile differences at different quantile levels
  simultaneously.



decr 0.0.7.9031
----------------------------------------------------------------

* Some dplyr wrappers used in various functions of the package.
* function dec_quantile() now computes also marginal quantiles 
  (reported in the last two rows of the output).
* function dec_() now reports also the total observed difference
  between the quantiles of the two groups (`delta_tot`), the difference
  observed in the common support (`delta_tot_CS`) and a "residual" part
  of the difference (`delta_AB`), which is the difference between
  `delta_tot` and `delta_tot_CS`.



decr 0.0.7.9030
----------------------------------------------------------------

* Removed 'ggplot2' in the list of suggested packages (DESCRIPTION).




decr 0.0.7.9029
----------------------------------------------------------------

* Use of 'roxygen2' 6.0.0



decr 0.0.7.9026
----------------------------------------------------------------

NEW FEATURES

* New generic function: nopodec_Fhat(). This estimates empirical cumulative
  distribution function, at a value `y`, for the two groups of observations,
  in and out the common support. In the common support, counterfactual ecdf
  are estimated. It also estimates the number of individuals of the two groups.
  For this generic, two S3 methods are available: nopodec_Fhat.default() and
  nopodec_Fhat.reweighted(.
* New generic function: margin_Fhat(). This has not yet any available S3
  method. The purpose of this function will be to estimate the observed
  empirical cumulative distribution function for groups A and B.





decr 0.0.7.9025
----------------------------------------------------------------

NEW FEATURES

* New function: margin_difference(). This takes the output of margin_mean() or
  margin_quantile() and simply computes the difference between the observed
  statistics of the two groups (always group A minus group B).
* Addedd various "See Also" sections in some functions of the package.



decr 0.0.7.9024
----------------------------------------------------------------

NEW FEATURES

* completed two (short) vignettes: `vignette("decr_intro")` and
  `vignette("bootstrap_example")`
* new function: reweight_strata_all4(), which produces an object of class
  `reweighted`.
* generic function of nopodec_mean(), dec_median(), dec_quantile(),
  margin_mean() and margin_quantile(). For each of these generics, two S3
  methods: default and reweighted.



decr 0.0.7.9021
----------------------------------------------------------------

NEW FEATURES

* Little modification to the reweight_strata_all2() function: added `y` and
  `variables` attributes to the output, that can be used in the next step.
* Added default value (`y = NULL`) in functions nopodec_mean(), dec_median(),
  dec_quantile(), margin_mean() and margin_quantile(). When invoked without
  specifying a value for `y`, the functions take the value from the previous
  step, the output of reweight_strata_all2().
* Completed the section "Value" of the documentation of the fhat_strata2()
  function.
* A new (invented) dataset has been added: `invented_wages`.


decr 0.0.7.9020
----------------------------------------------------------------

NEW FEATURES

* New functions: margin_mean() and margin_quantile().
* Completed the documentation of nopodec_mean() function (section Value)



decr 0.0.7.9019
----------------------------------------------------------------

NEW FEATURES

* Added a condition in the dec_() function to handle the case of no common
  support. If this is the case, the function prints a message and sets all the
  components equal to zero.


decr 0.0.7.9018
----------------------------------------------------------------

NEW FEATURES

* Solved the decomposition when group A members are all out of the common
  support, while only some of group B members are.

* Solved the decomposition when the "no out-of-support" case occurs, modifying
  the fhat_strata2() function.


decr 0.0.7.9017
----------------------------------------------------------------

NEW FEATURES

* New functions: boot_nopodec_mean() and boot_dec_quantile(). Both perform
  nonparametric stratified bootstrap of the average difference and quantile
  decompositions. These are the first functions for estimating confidence
  intervals of the various decomposition components. The bootstrap is performed
  with the 'boot' package.


decr 0.0.7.9016
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9015
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9014
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.90013
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9012
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9011
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9010
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9009
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9008
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9007
----------------------------------------------------------------

NEW FEATURES

* Minor modifications


decr 0.0.7.9006

R/nopodec.R


#' Decomposition of the observed difference between averages of y of group A and B in 4 components, as in Nopo (2008).
#'
#' The decomposition of the difference is like that developed by Nopo (2008), especially
#' regarding the two components out of support. The decomposition in the common support
#' is instead done with the reweighting approach proposed by Di Nardo, Fortin and Lemieux (1996).
#'
#' @param .nopodec_ output of \code{\link{nopodec_mean}}.
#' @param counterfactual "AB" or "BA". "AB" means that we want to estimate the counterfactual (wage) of group A, as if their characteristics were distributed as in group B. "BA" is the opposite (characteristics of group B are balanced to those of group A).
#'
#' @return a list with five components:
#'
#' \itemize{
#'   \item \code{delta_total}: total observed difference between average (wages)
#'    of group A and B;
#'   \item \code{delta_A}: part of the observed difference due to the fact that
#'   there are indidivuals of group A which are not comparable, for features,
#'    to individuals of group B;
#'   \item \code{delta_X} part explained by the fact that the two groups have a
#'    different distribution of characteristics;
#'   \item \code{delta_S} part not justified by the different distributions of the
#'    characteristics of the two groups, and potentially due to a difference
#'     in the remuneration structures between the two groups;
#'   \item \code{delta_B} part of the difference due to the fact that there are
#'    individuals of group B with characteristics that none of the group A
#'     has.
#' }
#'
#' @references
#' Nopo, H. 2008. Matching as a Tool to Decompose Wage Gaps. \emph{Review of Economics and Statistics}, 90(2): 290-299.
#'
#' DiNardo J., N. M. Fortin and T. Lemieux. 1996. Labor Market Institutions and the Distribution of Wages, 1973-1992: A Semiparametric Approach. \emph{Econometrica} 64 (5): 1001-44
#'
#'
#' @examples
#' data(invented_wages)
#'
#' # Common support and computation of counterfactual weights
#' r00 <- reweight_strata_all2(data = invented_wages,
#'                             treatment = "gender",
#'                             variables = c("sector", "education"),
#'                             y = "wage",
#'                             weights = "sample_weights")
#'
#' # Computation of the elements necessary to the decomposition
#' n00 <- nopodec_mean(r00)
#'
#' # Nopo decomposition
#' nopodec(n00, counterfactual = "AB")
#'
#' @export
nopodec <- function(.nopodec_, counterfactual = c("AB", "BA")){
  stopifnot(counterfactual[1] %in% c("AB", "BA"))

  treatment <- colnames(.nopodec_)[1]
  if(is.factor(.nopodec_[[treatment]])){
    groups_ <- levels(.nopodec_[[treatment]])
  }else{
    groups_ <- unique(.nopodec_[[treatment]])
  }

  # sel_A <- lazyeval::interp(~x == y, x = as.name(treatment), y = groups_[1])
  # sel_B <- lazyeval::interp(~x == y, x = as.name(treatment), y = groups_[2])
  sel_A <- paste0(treatment, " == ", "\"", groups_[1], "\"")
  sel_B <- paste0(treatment, " == ", "\"", groups_[2], "\"")

  # perc_ <- lazyeval::interp(~x / sum(x), x = as.name("Nhat"))
  Nhat <- rlang::sym("Nhat")
  Nhat <- rlang::enquo(Nhat)
  perc <- "perc"

  common_support <- "common_support"
  # sel_common_support <- lazyeval::interp(~!x, x = as.name(common_support))
  sel_common_support <- "!common_support"

  phat_AB_out <- .nopodec_ %>%
    gby_(treatment) %>%
    dplyr::mutate(!! perc := !! Nhat / sum(!! Nhat)) %>%
    # mutate2_(.dots = stats::setNames(list(perc_), c("perc"))) %>%
    filter2_(.dots = sel_common_support)

  phat_A_out <- (phat_AB_out %>% filter2_(sel_A))$perc
  phat_B_out <- (phat_AB_out %>% filter2_(sel_B))$perc

  ybar_A_out <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(!common_support))$ybar
  ybar_A_in <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(common_support))$ybar
  ybar_B_out <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(!common_support))$ybar
  ybar_B_in <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(common_support))$ybar

  # I due controfattuali (per una scomposizione ne serve uno dei due, ma la scomposizione
  # può essere fatta con entrambi: X_A'Beta_B e X_B'Beta_A)

  # Salario del gruppo A come se avesse le stesse caratteristiche del gruppo B (X_B'Beta_A)
  ybar_AB_C <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(common_support))$ybar_C_A

  # Salario del gruppo B come se avesse le stesse caratteristiche del gruppo A (X_A'Beta_B)
  ybar_BA_C <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(common_support))$ybar_C_B


  ################################################
  # Prova (per il caso no common support)
  if(check_numeric_0(ybar_A_in)) ybar_A_in <- 0
  if(check_numeric_0(ybar_B_in)) ybar_B_in <- 0
  if(check_numeric_0(ybar_AB_C)) ybar_AB_C <- 0
  if(check_numeric_0(ybar_BA_C)) ybar_BA_C <- 0
  ################################################

  # Le due componenti della scomposizione che si trovano fuori dal common support
  # non cambiano in funzione di quale controfattuale si usa (AB o BA)
  # N.B.: per "AB" si intende il controfattuale A come se le sue caratteristiche
  # fossero distribuite come quelle del gruppo B (Xbar_B'Beta_A in termini BO)
  delta_A <- phat_A_out * (ybar_A_out - ybar_A_in)
  delta_B <- phat_B_out * (ybar_B_in - ybar_B_out)

  counterfactual <- match.arg(counterfactual)
  # Prima scomposizione: con ybar_AB_C
  if(counterfactual == "AB"){
    delta_X <- ybar_A_in - ybar_AB_C
    delta_S <- ybar_AB_C - ybar_B_in
  }
  if(counterfactual == "BA"){
    delta_S <- ybar_A_in - ybar_BA_C
    delta_X <- ybar_BA_C - ybar_B_in
  }

  if(check_numeric_0(delta_A)) delta_A <- 0
  if(check_numeric_0(delta_B)) delta_B <- 0

  delta_tot <- delta_A + delta_X + delta_S + delta_B

  list(delta_tot = delta_tot, delta_A = delta_A, delta_X = delta_X,
       delta_S = delta_S, delta_B = delta_B)
}







nopodec_stat <- function(.nopodec_, counterfactual = c("AB", "BA"), stat = c("Fhat", "ybar")){
  stopifnot(counterfactual[1] %in% c("AB", "BA"))

  treatment <- colnames(.nopodec_)[1]
  if(is.factor(.nopodec_[[treatment]])){
    groups_ <- levels(.nopodec_[[treatment]])
  }else{
    groups_ <- unique(.nopodec_[[treatment]])
  }

  # sel_A <- lazyeval::interp(~x == y, x = as.name(treatment), y = groups_[1])
  # sel_B <- lazyeval::interp(~x == y, x = as.name(treatment), y = groups_[2])
  sel_A <- paste0(treatment, " == ", "\"", groups_[1], "\"")
  sel_B <- paste0(treatment, " == ", "\"", groups_[2], "\"")

  # perc_ <- lazyeval::interp(~x / sum(x), x = as.name("Nhat"))
  Nhat <- rlang::sym("Nhat")
  Nhat <- rlang::enquo(Nhat)
  perc <- "perc"

  common_support <- "common_support"
  # sel_common_support <- lazyeval::interp(~!x, x = as.name(common_support))
  sel_common_support <- "!common_support"

  phat_AB_out <- .nopodec_ %>%
    gby_(treatment) %>%
    dplyr::mutate(!! perc := !! Nhat / sum(!! Nhat)) %>%
    # mutate2_(.dots = stats::setNames(list(perc_), c("perc"))) %>%
    filter2_(.dots = sel_common_support)

  phat_A_out <- (phat_AB_out %>% filter2_(sel_A))$perc
  phat_B_out <- (phat_AB_out %>% filter2_(sel_B))$perc

  ybar_A_out <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(!common_support))[[stat[1]]]
  ybar_A_in <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(common_support))[[stat[1]]]
  ybar_B_out <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(!common_support))[[stat[1]]]
  ybar_B_in <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(common_support))[[stat[1]]]

  # I due controfattuali (per una scomposizione ne serve uno dei due, ma la scomposizione
  # può essere fatta con entrambi: X_A'Beta_B e X_B'Beta_A)

  stat_C_A <- paste0(stat[1], "_C_A")
  stat_C_B <- paste0(stat[1], "_C_B")
  # Salario del gruppo A come se avesse le stesse caratteristiche del gruppo B (X_B'Beta_A)
  ybar_AB_C <- (.nopodec_ %>% filter2_(sel_A) %>% dplyr::filter(common_support))[[stat_C_A]]

  # Salario del gruppo B come se avesse le stesse caratteristiche del gruppo A (X_A'Beta_B)
  ybar_BA_C <- (.nopodec_ %>% filter2_(sel_B) %>% dplyr::filter(common_support))[[stat_C_B]]


  ################################################
  # Prova (per il caso no common support)
  if(check_numeric_0(ybar_A_in)) ybar_A_in <- 0
  if(check_numeric_0(ybar_B_in)) ybar_B_in <- 0
  if(check_numeric_0(ybar_AB_C)) ybar_AB_C <- 0
  if(check_numeric_0(ybar_BA_C)) ybar_BA_C <- 0
  ################################################

  # Le due componenti della scomposizione che si trovano fuori dal common support
  # non cambiano in funzione di quale controfattuale si usa (AB o BA)
  # N.B.: per "AB" si intende il controfattuale A come se le sue caratteristiche
  # fossero distribuite come quelle del gruppo B (Xbar_B'Beta_A in termini BO)
  delta_A <- phat_A_out * (ybar_A_out - ybar_A_in)
  delta_B <- phat_B_out * (ybar_B_in - ybar_B_out)

  counterfactual <- match.arg(counterfactual)
  # Prima scomposizione: con ybar_AB_C
  if(counterfactual == "AB"){
    delta_X <- ybar_A_in - ybar_AB_C
    delta_S <- ybar_AB_C - ybar_B_in
  }
  if(counterfactual == "BA"){
    delta_S <- ybar_A_in - ybar_BA_C
    delta_X <- ybar_BA_C - ybar_B_in
  }

  if(check_numeric_0(delta_A)) delta_A <- 0
  if(check_numeric_0(delta_B)) delta_B <- 0

  delta_tot <- delta_A + delta_X + delta_S + delta_B

  list(delta_tot = delta_tot, delta_A = delta_A, delta_X = delta_X,
       delta_S = delta_S, delta_B = delta_B)
}
